torch/csrc/jit/passes/common_expression_hoisting.cpp

```diff
@@ -0,0 +1,142 @@
+#include <torch/csrc/jit/passes/common_expression_hoisting.h>
+
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/node_hashing.h>
+#include <torch/csrc/jit/jit_log.h>
+
+#include <cstddef>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace {
+
+struct CommonExpressionHoister {
+  CommonExpressionHoister(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    return HoistCommonExpression(graph_->block());
+  }
+
+  bool HoistFromIfNode(Node* if_node) {
+    Block* true_block = if_node->blocks()[0];
+    Block* false_block = if_node->blocks()[1];
+    // find common statements in the two subblocks
+    bool block_modified = false;
+
+    auto true_block_nodes = std::unordered_set<Node*, HashNode, EqualNode>(
+        true_block->nodes().begin(), true_block->nodes().end());
+    for (auto it = false_block->nodes().begin();
+         it != false_block->nodes().end();) {
+      Node* false_b_node = *it;
+      // node may be moved to a different block so advance iterator now
+      ++it;
+
+      auto matching_elem = true_block_nodes.find(false_b_node);
+      if (matching_elem == true_block_nodes.end()) {
+        continue;
+      }
+      Node* true_b_node = *matching_elem;
+
+      // Check if a move to the front of the block is valid
+      // If both of the moves are valid, then we know we can move the item out
+      // of the if blocks entirely.
+      AliasDb& aliasDb = getOrCreateAliasDb();
+      bool true_moveable = aliasDb.couldMoveAfterTopologically(
+          true_b_node, true_block->nodes().front());
+      bool false_moveable = aliasDb.couldMoveAfterTopologically(
+          false_b_node, false_block->nodes().front());
+
+      if (!true_moveable || !false_moveable) {
+        continue;
+      }
+
+      // Get all the uses of the output to delete and reinsert them
+      // as the input would change, the HashNode value would also change.
+      std::unordered_set<Node*> true_b_uses;
+      for (Value* true_out : true_b_node->outputs()) {
+        for (Use true_use : true_out->uses()) {
+          true_b_uses.insert(true_use.user);
```

Comment:
we also want to check that the uses are on the right block, you can check `owningBlock()`

Reply:
Why would the use of the variable not be in that block or a compatible outer block?

Reply:
```
prim::If
    block1():
          a = op()
          op1(a)
          prim::If
                 block2():
                        op2(a)
```

We don't want to add recursively contained uses, e.g. we just want uses on block1, not block2. The nodes in block2 are not meant to be in the true_block_nodes